File: finalAssignmentBackend/src/main/java/com/tutict/finalassignmentbackend/config/security/pqc/MlDsaKeyRing.java
package com.tutict.finalassignmentbackend.config.security.pqc;

import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
import org.bouncycastle.cert.X509CertificateHolder;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.openssl.PEMParser;
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;

import java.io.StringReader;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * 版本化 ML-DSA 密钥环（与 cloud 端一致）：按 {@code kid} 存储多把密钥，支持"新旧密钥双验"的在线轮换。
 *
 * <ul>
 *   <li>签名方使用 {@link #activeEntry()}（当前活跃密钥）签名，并在 JWT header 写入 kid；</li>
 *   <li>校验方通过 {@link #publicKeyFor(String)} 按 kid 选择公钥；token 无 kid 或 kid 未知时
 *       回退到 {@link #activePublicKey()}；</li>
 *   <li>{@link #activate(String, PrivateKey, PublicKey)} 在运行时生成新密钥并切换活跃密钥，
 *       旧密钥保留在环中供校验，直到 {@link #retireOlderThan(Duration)} 清理。</li>
 * </ul>
 */
public final class MlDsaKeyRing {

    private static final Logger LOG = Logger.getLogger(MlDsaKeyRing.class.getName());
    private static final String BC = BouncyCastleProvider.PROVIDER_NAME;
    public static final String ML_DSA_ALGORITHM = "ML-DSA-65";
    public static final String EPHEMERAL_KID = "ml-dsa-ephemeral";

    private final ConcurrentMap<String, MlDsaKeyEntry> entries = new ConcurrentHashMap<>();
    private volatile String activeKid;

    private MlDsaKeyRing() {
    }

    /** 从配置 + 遗留单钥回退构建密钥环。配置 keys 非空时优先；否则尝试单钥；再否则生成临时密钥。 */
    public static MlDsaKeyRing from(MlDsaKeyRingProperties props, String legacyPrivateKeyPem, String legacyPublicKeyPem) {
        PqcProviderInitializer.ensureBouncyCastle();
        MlDsaKeyRing ring = new MlDsaKeyRing();
        List<MlDsaKeyProperties> keys = props != null ? props.getKeys() : List.of();
        if (keys != null && !keys.isEmpty()) {
            for (MlDsaKeyProperties key : keys) {
                if (key == null || !isPresent(key.getKid()) || !isPresent(key.getPublicKey())) {
                    continue;
                }
                try {
                    PrivateKey privateKey = isPresent(key.getPrivateKey()) ? loadPrivateKey(key.getPrivateKey()) : null;
                    PublicKey publicKey = loadPublicKey(key.getPublicKey());
                    ring.add(key.getKid(), privateKey, publicKey);
                } catch (Exception ex) {
                    LOG.log(Level.WARNING, "Failed to load configured ML-DSA key kid=" + key.getKid(), ex);
                }
            }
            String preferred = props != null ? props.getActiveKid() : null;
            if (isPresent(preferred) && ring.entries.containsKey(preferred)) {
                ring.activeKid = preferred;
            } else {
                ring.activeKid = ring.pickActiveKid();
            }
            return ring;
        }
        if (isPresent(legacyPrivateKeyPem) && isPresent(legacyPublicKeyPem)) {
            try {
                ring.add("current", loadPrivateKey(legacyPrivateKeyPem), loadPublicKey(legacyPublicKeyPem));
                ring.activeKid = "current";
                return ring;
            } catch (Exception ex) {
                LOG.log(Level.WARNING, "Failed to load legacy ML-DSA keys, generating ephemeral pair", ex);
            }
        }
        // Ephemeral: tokens will NOT survive a restart.
        try {
            KeyPairGenerator kpg = KeyPairGenerator.getInstance(ML_DSA_ALGORITHM, BC);
            KeyPair kp = kpg.generateKeyPair();
            ring.add(EPHEMERAL_KID, kp.getPrivate(), kp.getPublic());
            ring.activeKid = EPHEMERAL_KID;
            LOG.warning("No ML-DSA keys configured; generated ephemeral keypair (kid=" + EPHEMERAL_KID
                    + "). Tokens will NOT survive a restart.");
        } catch (Exception ex) {
            throw new IllegalStateException("Failed to generate ephemeral ML-DSA keypair", ex);
        }
        return ring;
    }

    public boolean containsKid(String kid) {
        return kid != null && entries.containsKey(kid);
    }

    public Optional<PublicKey> publicKeyFor(String kid) {
        if (!isPresent(kid)) {
            // 无 kid 的历史 token：回退到当前活跃公钥。
            return Optional.ofNullable(activePublicKey());
        }
        MlDsaKeyEntry entry = entries.get(kid);
        return entry == null ? Optional.empty() : Optional.ofNullable(entry.publicKey());
    }

    public PublicKey activePublicKey() {
        MlDsaKeyEntry active = entries.get(activeKid);
        return active != null ? active.publicKey() : null;
    }

    public PrivateKey activePrivateKey() {
        MlDsaKeyEntry active = entries.get(activeKid);
        return active != null ? active.privateKey() : null;
    }

    public String activeKid() {
        return activeKid;
    }

    public Set<String> kidSet() {
        return entries.keySet();
    }

    public int size() {
        return entries.size();
    }

    /**
     * 轮换：生成/注入新密钥并切换为活跃签名密钥。旧密钥保留在环中继续供校验。
     */
    public synchronized void activate(String kid, PrivateKey privateKey, PublicKey publicKey) {
        add(kid, privateKey, publicKey);
        this.activeKid = kid;
        LOG.info(() -> "ML-DSA active signing key switched to kid=" + kid
                + ", ring size=" + entries.size());
    }

    /** 移除非活跃且激活时间早于 retention 的旧密钥（双验窗口结束后的清理）。 */
    public synchronized void retireOlderThan(Duration retention) {
        Instant cutoff = Instant.now().minus(retention);
        int removed = 0;
        for (String kid : entries.keySet()) {
            if (kid.equals(activeKid)) {
                continue;
            }
            MlDsaKeyEntry entry = entries.get(kid);
            if (entry != null && entry.activatedAt().isBefore(cutoff)) {
                entries.remove(kid);
                removed++;
            }
        }
        if (removed > 0) {
            final int removedCount = removed;
            LOG.info(() -> "Retired " + removedCount + " expired ML-DSA verification key(s); active kid=" + activeKid);
        }
    }

    private void add(String kid, PrivateKey privateKey, PublicKey publicKey) {
        entries.put(kid, new MlDsaKeyEntry(kid, privateKey, publicKey, Instant.now()));
    }

    private String pickActiveKid() {
        String withPrivate = null;
        String first = null;
        for (MlDsaKeyEntry entry : entries.values()) {
            if (first == null) {
                first = entry.kid();
            }
            if (entry.privateKey() != null) {
                withPrivate = entry.kid();
            }
        }
        return withPrivate != null ? withPrivate : first;
    }

    private static PrivateKey loadPrivateKey(String pem) throws Exception {
        try (PEMParser parser = new PEMParser(new StringReader(pem))) {
            Object obj = parser.readObject();
            if (obj instanceof PrivateKeyInfo pki) {
                return new JcaPEMKeyConverter().setProvider(BC).getPrivateKey(pki);
            }
            throw new IllegalArgumentException("PEM is not a PKCS#8 private key: " + obj);
        }
    }

    private static PublicKey loadPublicKey(String pem) throws Exception {
        try (PEMParser parser = new PEMParser(new StringReader(pem))) {
            Object obj = parser.readObject();
            JcaPEMKeyConverter conv = new JcaPEMKeyConverter().setProvider(BC);
            if (obj instanceof SubjectPublicKeyInfo spki) {
                return conv.getPublicKey(spki);
            }
            if (obj instanceof X509CertificateHolder cert) {
                return conv.getPublicKey(cert.getSubjectPublicKeyInfo());
            }
            throw new IllegalArgumentException("PEM is not a public key: " + obj);
        }
    }

    private static boolean isPresent(String s) {
        return s != null && !s.isBlank();
    }

    /**
     * 环中的一个版本条目。privateKey 可能为 null（仅校验方）。
     */
    public record MlDsaKeyEntry(String kid, PrivateKey privateKey, PublicKey publicKey, Instant activatedAt) {
    }
}

File: final_assignment_backend_go/project/configs/auth/mldsa_keyring.go
// ML-DSA-65（FIPS 204）版本化密钥环，对齐 Spring/Quarkus 后端的
// config/security/pqc/MlDsaKeyRing 语义：按 kid 存储多把密钥，支持"新旧密钥双验"的在线轮换。
//
// 注意：Go 端使用 cloudflare/circl 的 ML-DSA-65 实现，密钥为 FIPS 204 原始打包字节
// （PublicKeySize / PrivateKeySize 固定长度），通过 base64 编码配置；
// 这与 Spring/BC 的 PKCS#8 PEM 格式不同，两端不直接互通密钥，
// 但各后端独立运行时具备对等的密钥环/轮换/kid 能力。
package auth

import (
	"crypto/rand"
	"encoding/base64"
	"errors"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/cloudflare/circl/sign/mldsa/mldsa65"
)

// ML-DSA-65 相关常量，对齐 Spring MlDsaKeyRing。
const (
	MlDsaAlgorithm = "ML-DSA-65"
	MlDsaJwtAlg    = "ML-DSA-65" // JWT header alg
	EphemeralKid   = "ml-dsa-ephemeral"
)

// MlDsaKeyEntry 密钥环中的一个版本条目。PrivateKey 可能为 nil（仅校验方）。
type MlDsaKeyEntry struct {
	Kid         string
	PrivateKey  *mldsa65.PrivateKey
	PublicKey   *mldsa65.PublicKey
	ActivatedAt time.Time
}

// MlDsaKeyRing 版本化 ML-DSA 密钥环。线程安全。
type MlDsaKeyRing struct {
	mu        sync.RWMutex
	entries   map[string]MlDsaKeyEntry
	activeKid string
}

// KeyConfig 单个版本化密钥的配置项（base64 编码的原始打包字节）。
type KeyConfig struct {
	Kid        string
	PublicKey  string // base64
	PrivateKey string // base64；校验方可空
}

// NewMlDsaKeyRing 从配置构建密钥环。
//   - keys 非空时优先使用版本化密钥环，activeKid 缺省则取最后一个含私钥的条目；
//   - 否则若 legacy 公/私钥均提供，使用单钥（kid="current"）；
//   - 否则生成临时密钥（重启后历史 token 失效）。
func NewMlDsaKeyRing(activeKid string, keys []KeyConfig, legacyPublicB64, legacyPrivateB64 string) (*MlDsaKeyRing, error) {
	ring := &MlDsaKeyRing{entries: map[string]MlDsaKeyEntry{}}

	if len(keys) > 0 {
		for _, k := range keys {
			if k.Kid == "" || k.PublicKey == "" {
				continue
			}
			pk, err := decodePublicKey(k.PublicKey)
			if err != nil {
				log.Printf("[WARN] Failed to load configured ML-DSA public key kid=%s: %v", k.Kid, err)
				continue
			}
			var sk *mldsa65.PrivateKey
			if k.PrivateKey != "" {
				if sk, err = decodePrivateKey(k.PrivateKey); err != nil {
					log.Printf("[WARN] Failed to load configured ML-DSA private key kid=%s: %v", k.Kid, err)
				}
			}
			ring.entries[k.Kid] = MlDsaKeyEntry{Kid: k.Kid, PrivateKey: sk, PublicKey: pk, ActivatedAt: time.Now()}
		}
		if activeKid != "" {
			if _, ok := ring.entries[activeKid]; ok {
				ring.activeKid = activeKid
			} else {
				ring.activeKid = ring.pickActiveKid()
			}
		} else {
			ring.activeKid = ring.pickActiveKid()
		}
		if ring.activeKid == "" {
			return nil, errors.New("ML-DSA key ring has no usable keys")
		}
		return ring, nil
	}

	if legacyPublicB64 != "" && legacyPrivateB64 != "" {
		pk, err := decodePublicKey(legacyPublicB64)
		if err != nil {
			return nil, fmt.Errorf("failed to decode legacy ML-DSA public key: %w", err)
		}
		sk, err := decodePrivateKey(legacyPrivateB64)
		if err != nil {
			return nil, fmt.Errorf("failed to decode legacy ML-DSA private key: %w", err)
		}
		ring.entries["current"] = MlDsaKeyEntry{Kid: "current", PrivateKey: sk, PublicKey: pk, ActivatedAt: time.Now()}
		ring.activeKid = "current"
		return ring, nil
	}

	// Ephemeral: tokens will NOT survive a restart.
	pk, sk, err := mldsa65.GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("failed to generate ephemeral ML-DSA keypair: %w", err)
	}
	ring.entries[EphemeralKid] = MlDsaKeyEntry{Kid: EphemeralKid, PrivateKey: sk, PublicKey: pk, ActivatedAt: time.Now()}
	ring.activeKid = EphemeralKid
	log.Println("[WARN] No ML-DSA keys configured; generated ephemeral keypair (kid=" + EphemeralKid +
		"). Tokens will NOT survive a restart.")
	return ring, nil
}

// ActiveKid 返回当前活跃签名密钥的 kid。
func (r *MlDsaKeyRing) ActiveKid() string {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.activeKid
}

// ActivePrivateKey 返回活跃私钥；不存在则 nil。
func (r *MlDsaKeyRing) ActivePrivateKey() *mldsa65.PrivateKey {
	r.mu.RLock()
	defer r.mu.RUnlock()
	e, ok := r.entries[r.activeKid]
	if !ok {
		return nil
	}
	return e.PrivateKey
}

// ActivePublicKey 返回活跃公钥；不存在则 nil。
func (r *MlDsaKeyRing) ActivePublicKey() *mldsa65.PublicKey {
	r.mu.RLock()
	defer r.mu.RUnlock()
	e, ok := r.entries[r.activeKid]
	if !ok {
		return nil
	}
	return e.PublicKey
}

// PublicKeyFor 按 kid 选择公钥。kid 为空或未知时回退到活跃公钥（对齐 Spring 语义）。
func (r *MlDsaKeyRing) PublicKeyFor(kid string) (*mldsa65.PublicKey, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	if kid == "" {
		if e, ok := r.entries[r.activeKid]; ok {
			return e.PublicKey, nil
		}
		return nil, errors.New("no active ML-DSA verification key")
	}
	e, ok := r.entries[kid]
	if !ok {
		// 未知 kid 回退到活跃公钥，兼容轮换后旧 token 的边界场景
		if ae, ok := r.entries[r.activeKid]; ok {
			return ae.PublicKey, nil
		}
		return nil, fmt.Errorf("no verification key for ML-DSA kid=%s", kid)
	}
	return e.PublicKey, nil
}

// Activate 注入新密钥并切换为活跃签名密钥。旧密钥保留在环中继续供校验。
func (r *MlDsaKeyRing) Activate(kid string, sk *mldsa65.PrivateKey, pk *mldsa65.PublicKey) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.entries[kid] = MlDsaKeyEntry{Kid: kid, PrivateKey: sk, PublicKey: pk, ActivatedAt: time.Now()}
	r.activeKid = kid
	log.Printf("[INFO] ML-DSA active signing key switched to kid=%s, ring size=%d", kid, len(r.entries))
}

// RetireOlderThan 清理激活时间早于 retention 的非活跃旧密钥（双验窗口结束后的清理）。
func (r *MlDsaKeyRing) RetireOlderThan(retention time.Duration) {
	r.mu.Lock()
	defer r.mu.Unlock()
	cutoff := time.Now().Add(-retention)
	removed := 0
	for kid, e := range r.entries {
		if kid == r.activeKid {
			continue
		}
		if e.ActivatedAt.Before(cutoff) {
			delete(r.entries, kid)
			removed++
		}
	}
	if removed > 0 {
		log.Printf("[INFO] Retired %d expired ML-DSA verification key(s); active kid=%s", removed, r.activeKid)
	}
}

// Size 返回环中密钥数量。
func (r *MlDsaKeyRing) Size() int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return len(r.entries)
}

// Rotate 生成新 ML-DSA-65 密钥对，以时间戳 kid 注入并切换为活跃签名密钥，
// 然后清理超过 retention 的旧密钥。返回新 kid 与需要分发给校验方的新公钥（base64）。
func (r *MlDsaKeyRing) Rotate(retention time.Duration) (kid string, publicKeyB64 string, err error) {
	pk, sk, err := mldsa65.GenerateKey(rand.Reader)
	if err != nil {
		return "", "", fmt.Errorf("failed to generate ML-DSA keypair: %w", err)
	}
	// 毫秒精度，避免同秒内多次轮换（定时 + 手动触发，或多副本同窗口）生成相同 kid
	// 并互相覆盖，导致前一次窗口内签发的 token 解析到错误公钥而验签失败。
	kid = "ml-dsa-" + time.Now().UTC().Format("20060102-150405.000")
	r.Activate(kid, sk, pk)
	r.RetireOlderThan(retention)
	publicKeyB64 = base64.StdEncoding.EncodeToString(pk.Bytes())
	log.Printf("[INFO] ML-DSA key rotated to kid=%s (ring size=%d). Distribute the new public key to verifiers.", kid, r.Size())
	return kid, publicKeyB64, nil
}

func (r *MlDsaKeyRing) pickActiveKid() string {
	withPrivate := ""
	first := ""
	for _, e := range r.entries {
		if first == "" {
			first = e.Kid
		}
		if e.PrivateKey != nil {
			withPrivate = e.Kid
		}
	}
	if withPrivate != "" {
		return withPrivate
	}
	return first
}

func decodePublicKey(b64 string) (*mldsa65.PublicKey, error) {
	raw, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		return nil, fmt.Errorf("base64 decode public key: %w", err)
	}
	pk := &mldsa65.PublicKey{}
	if err := pk.UnmarshalBinary(raw); err != nil {
		return nil, fmt.Errorf("unmarshal public key: %w", err)
	}
	return pk, nil
}

func decodePrivateKey(b64 string) (*mldsa65.PrivateKey, error) {
	raw, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		return nil, fmt.Errorf("base64 decode private key: %w", err)
	}
	sk := &mldsa65.PrivateKey{}
	if err := sk.UnmarshalBinary(raw); err != nil {
		return nil, fmt.Errorf("unmarshal private key: %w", err)
	}
	return sk, nil
}
